evaluation/functions.py

```diff
@@ -0,0 +1,125 @@
+        url = f"{base_url}/conversation"
+        headers = {
+            "content-type": "application/json",
+        }
+
+        response = requests.post(url, json=data, headers=headers)
+
+        response.raise_for_status()  # Raises HTTPError for bad responses
+
+        response_json = response.json()
+
+        return response_json
+
+    except requests.exceptions.RequestException as e:
+        raise Exception(f"Error in making the API request.\n{e}")
+
+    except (KeyError, IndexError) as e:
+        raise Exception(f"Error in parsing the API response.\n{e}")
+
+
+def process_prompt(prompt: str):
+    """
+    Process a row of data containing a user question and obtain the AI-generated response along with an evaluation response/rating.
+
+    Parameters:
+    - prompt (string): A string containig Prmopt data.
+
+    """
+    try:
+        prompt_data = {"messages": [{"role": "user", "content": prompt}]}
+        start_time = time.time()
+        response = get_response(prompt_data)
+        end_time = time.time()
+        answer = response["choices"][0]["messages"][1]["content"]
+        response_time = end_time - start_time
+        return answer, response_time, response
+
+    except (KeyError, IndexError) as e:
+        raise Exception(f"Error in parsing the API response.\n{e}")
+
+
+# Function to rate the answer
+def evaluate_response(prompt: str):
+    """
+    Rate the quality of an AI-generated response based on the provided response and original user question.
+
+    Parameters:
+    - prompt (str): The original user question.
+    """
+    answer, response_time, response = process_prompt(prompt)
+    evaluation_string = os.environ.get(
+        "PROMPT_EVAL_STRING", "Evaluate the above response quality"
+    )
+    try:
+        prompt_data = {
+            "messages": [
+                {
+                    "role": "user",
+                    "content": prompt,
+                },
+                {
+                    "role": "tool",
+                    "content": response["choices"][0]["messages"][0]["content"],
+                },
+                {
+                    "role": "assistant",
+                    "content": response["choices"][0]["messages"][1]["content"],
+                },
+                {"role": "user", "content": evaluation_string},
+            ]
+        }
+        response = get_response(prompt_data)
+        evaluation = response["choices"][0]["messages"][1]["content"]
+        return prompt, answer, response_time, evaluation
+
+    except (KeyError, IndexError) as e:
+        raise Exception(f"Error in parsing the API response.\n{e}")
+
+
+def evaluate_and_compare_prompt_responses(prompt: str):
+    """
+    Evaluate and compare responses for a given prompt.
+
+    Parameters:
+    - prompt (str): The user prompt to be processed and compared.
+    """
+    try:
+        answer_1, response_time_1, _ = process_prompt(prompt)
```

Comment:
This appears to be running the same AI pipeline/orchestrator twice on the same input prompt. And since the temperature is 0 the responses will most likely be the same or very close. Is there a plan to make the orchestrator configurable so we can compare the results of changing the system prompt and/or search strategy etc.?

Reply:
Will rewrite the script to do comparison with answers from the previous run and not within the same run/session.

Reply:
It will still be the same AI orchestrator/pipeline though, so the answers to compare will be very close if not identical. The real value in the comparison is to be able to evaluate against differences in the pipeline itself, e.g., a different prompt, a different search strategy, parameter tweaks, etc.